src/autocast_fp8.cpp

```diff
@@ -0,0 +1,87 @@
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+#include <migraphx/autocast_fp8.hpp>
+#include <migraphx/make_op.hpp>
+#include <migraphx/program.hpp>
+#include <migraphx/instruction.hpp>
+#include <migraphx/iterator_for.hpp>
+
+namespace migraphx {
+inline namespace MIGRAPHX_INLINE_NS {
+
+bool autocast_fp8_pass::is_fp8_type(const shape::type_t& s) const
+{
+    return fp8_types.find(s) != fp8_types.end();
+}
+
+void autocast_fp8_pass::apply(module& m) const
+{
+    std::vector<instruction_ref> remove_parameters;
+    for(auto ins : iterator_for(m))
+    {
+        const auto& ins_name = ins->name();
+        if(ins_name == "@param" and is_fp8_type(ins->get_shape().type()))
+        {
+            shape::type_t fp8_type = ins->get_shape().type();
+            migraphx::shape new_shape = ins->get_shape().with_type(target_type);
+            std::string new_param_name =
+                ins->get_operator().to_value()["parameter"].to<std::string>();
+            auto new_param = m.add_parameter(new_param_name, new_shape);
+            auto new_ins = m.insert_instruction(
+                ins,
+                migraphx::make_op("convert", {{"target_type", migraphx::to_value(fp8_type)}}), new_param);
+            m.replace_instruction(ins, new_ins);
+            remove_parameters.push_back(ins);
+        }
+
+        if(ins_name == "@return")
+        {
+            std::vector<instruction_ref> inputs = ins->inputs();
+            std::vector<instruction_ref> new_inputs;
+            for(const auto& i : inputs)
+            {
+                if (is_fp8_type(i->get_shape().type()))
+                {
+                    new_inputs.push_back(m.insert_instruction(
+                        ins,
+                        migraphx::make_op("convert", {{"target_type", migraphx::to_value(target_type)}}),
+                        i));
+                }
+            }
+            if(new_inputs.size())
```

Comment:
this won't work. new_inputs would only have converted instructions. 

if `return` is let's say returning mixed types 
fp32 and fp8 then new_inputs would only have fp8 one. 

also there could be multiple elements inside `new_inputs` vector.

Reply:
You are right, thanks for catching that. I have made the required changes.